adam_core/utils/indexable.py

```diff
@@ -1,174 +1,446 @@
+        )
 
-        elif isinstance(index, np.ndarray):
-            class_index = index
-            self._index_attribute = None
-            array_index = np.arange(0, len(index), dtype=int)
+    @class_index_to_members.deleter
+    def class_index_to_members_is_slice(self):
+        raise ValueError(
+            "Cannot delete class_index_to_members_is_slice. Use set_index() with no arguments instead."
+        )
 
-        else:
-            err = "index must be a str, numpy.ndarray"
-            raise ValueError(err)
+    @property
+    def class_index_attribute(self):
+        """
+        Attribute on which the class index was set.
+        """
+        return self._class_index_attribute
 
-        self._index = pd.MultiIndex.from_arrays(
-            [class_index, array_index], names=["class_index", "array_index"]
+    @class_index_attribute.setter
+    def class_index_attribute(self, value):
+        raise ValueError(
+            "Cannot set class_index_attribute directly. Use set_index() instead."
         )
 
-        return
+    @class_index_attribute.deleter
+    def class_index_attribute(self):
+        raise ValueError(
+            "Cannot delete class_index_attribute. Use set_index() with no arguments instead."
+        )
 
-    def __getitem__(self, i: Union[int, slice, tuple, list, np.ndarray, pd.MultiIndex]):
-        ind = self._handle_index(i)
-        copy = deepcopy(self)
+    @property
+    def member_index(self):
+        """
+        Index of the class members.
+        """
+        return self._member_index
 
-        for k, v in self.__dict__.items():
-            if isinstance(
-                v,
-                (np.ndarray, np.ma.masked_array, list, Time, Indexable, pd.MultiIndex),
-            ):
-                copy.__dict__[k] = v[ind]
-            elif isinstance(v, UNSLICEABLE_DATA_STRUCTURES):
-                copy.__dict__[k] = v
-            elif v is None:
-                pass
-            else:
-                err = f"{type(v)} are not supported."
-                raise NotImplementedError(err)
+    @member_index.setter
+    def member_index(self, value):
+        raise ValueError("Cannot set member_index directly. Use set_index() instead.")
 
-        return copy
+    @member_index.deleter
+    def member_index(self):
+        raise ValueError(
+            "Cannot delete member_index. Use set_index() with no arguments instead."
+        )
 
-    def __delitem__(self, i: Union[int, slice, tuple, list, np.ndarray, pd.MultiIndex]):
-        ind = self._handle_index(i)
+    @property
+    def member_length(self):
+        """
+        Length of the class members.
+        """
+        return self._member_length
+
+    @member_length.setter
+    def member_length(self, value):
+        raise ValueError("Cannot set member_length directly. Use set_index() instead.")
 
+    @member_length.deleter
+    def member_length(self):
```

Comment:
My comments above on the `class_index` property apply equally to `class_index_to_members`, `class_index_to_members_is_slice`, `class_index_attribute`, `member_index`, and `member_length`. I would delete all of those.

Reply:
Fixed.